### PR TITLE
Add diffing logic

### PR DESCRIPTION
**Pull request by sdlyy**
Resolves L2B-2867

## What's changed

- Added changelog entries struct
- Added discriminated modification types
- Very base has been migrated from discovery diff with few changes
- Only `values` are a subject to be diffed
- We can expand changelog entries to code, ABIs and root properties by assigning yet another changelog entry's property `family` (example one) that would indicate other group given diff entry belongs to
- No pretty diffs are preserved - only raw ones. We can add it, however in case of nested objects/arrays it will take reasonable amount of space (we probably don't care about it much tho)
- Contracts are matched and split into safe pairs 

## How to test it

Feel free to insert your own discovery outputs into test cases of `diff.test.ts`

## Notes

- <strike>What if contract appears/disappears between diffs? Currently, pair-matching is naive and treats the previous output as a reference - if the contract was present within the first output and is not present within the second one, then it is marked as unsafe and skipped for now. If contract is not present within first output and present within second one, it is discarded completely (not even included in unsafe pairs)</strike> Will be part of changelog indexer

### CODE REVIEW

File: packages/backend/src/tools/changelog/pairs.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+type UnsafeContractPair = readonly [
+  ContractParameters,
+  ContractParameters | undefined,
+]
+
+type SafeContractPair = readonly [ContractParameters, ContractParameters]
+
+// What if given contract appeared between previous and current?
+// What if given contract disappeared between previous and current?
```
Comment (michalsidzej) [inline packages/backend/src/tools/changelog/pairs.ts:13]: We should know about it in the changelog. Both of those cases are already in our discovery DB table.

File: packages/backend/src/tools/changelog/diff.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  getDiscoveryChanges,
+}
+
+function getDiscoveryChanges(
+  previousOutput: DiscoveryOutput,
+  currentOutput: DiscoveryOutput,
+): ChangelogEntry[] {
+  const matchingPairs = getMatchingContractPairs(previousOutput, currentOutput)
+
+  const { safe: safePairs } = splitSafeContractPairs(matchingPairs)
```
Comment (michalsidzej) [inline packages/backend/src/tools/changelog/diff.ts:22]: So we should do something with unsafe pairs. We should store information that contract appeared/disappeared from the system.

**Review by michalsidzej — COMMENTED** [review]
LGTM. The only thing left is to make sure we mark the apperance/disappearance of contracts

File: packages/backend/src/tools/changelog/diff.ts (hunk 2 above)
Comment (sdlyy) [reply packages/backend/src/tools/changelog/diff.ts]: Will be a part of changelog indexer - already implemented on draft PR

File: packages/backend/src/tools/changelog/pairs.ts (hunk 1 above)
Comment (sdlyy) [reply packages/backend/src/tools/changelog/pairs.ts]: Will be a part of changelog indexer - already implemented on draft PR

File: packages/backend/src/tools/changelog/diff.test.ts [hunk 3]
```diff
@@ -0,0 +1,634 @@
...
+        parameterPath: ['defaultAdapterParams', '101', '1'],
+        previousValue: null,
+        currentValue:
+          '{"proofType":2,"adapterParams":"0x00010000000000000000000000000000000000000000000000000000000000030d48"}',
+      },
+    ])
+  })
+
+  it('handles contract creation', () => {
+    const currentEndpoint = {
```
Comment (michalsidzej) [inline packages/backend/src/tools/changelog/diff.test.ts:156]: Just a suggestion: a case with the empty contract (no values) could be added for docs purposes

**Review by michalsidzej — APPROVED** [review]

**Pull request merged**